src/ppx-metaquot/fun_ty.mli

```diff
@@ -0,0 +1,143 @@
+(* This file is part of Learn-OCaml.
+ *
+ * Copyright (C) 2019 OCaml Software Foundation.
+ * Copyright (C) 2016-2018 OCamlPro.
+ *
+ * Learn-OCaml is distributed under the terms of the MIT license. See the
+ * included LICENSE file for details. *)
+
+(** [Fun_ty] is used by [Test_lib] to implement n-ary graders.
+
+    This module provides two GADTs: [args], representing arguments of
+    n-ary functions as nonempty heterogeneous lists, and [fun_ty],
+    representing function types in a more explicit way than ['a Ty.ty].
+
+    This module also serves as a runtime library for the customized
+    [ppx_metaquot], so that one may write:
+
+    [let p = [%funty: int -> bool]], and get:
+
+    [val p : ((int -> bool) Ty.ty, int -> unit, bool) fun_ty = <abstr>] *)
+
+(** GADT for arguments of n-ary functions, implemented as nonempty
+    heterogeneous lists.
+
+    Usage: [arg 3 @@ arg "word" @@ last false]
+
+    Alternatively: [3 @: "word" @:!! false] *)
+type ('arrow, 'uarrow, 'ret) args
+
+(** [last e], or equivalently [!! e], builds a one-element argument list *)
+val last :
+  'a ->
+  ('a -> 'ret, 'a -> unit, 'ret) args
+
+(** [arg a l], or equivalently [a @: l], adds [a] in front of the
+   argument list [l] *)
+val arg :
+  'a ->
+  ('ar -> 'row, 'ar -> 'urow, 'ret) args ->
+  ('a -> 'ar -> 'row, 'a -> 'ar -> 'urow, 'ret) args
+
+(** Helper notation for [last] *)
+val (!!) :
+  'a ->
+  ('a -> 'ret, 'a -> unit, 'ret) args
+
+(** Helper notation for [arg] *)
+val (@:) :
+  'a ->
+  ('ar -> 'row, 'ar -> 'urow, 'ret) args ->
+  ('a -> 'ar -> 'row, 'a -> 'ar -> 'urow, 'ret) args
+
+(** [a @:!! l] is another notation for [a @: !! l] (with a space) *)
+val (@:!!) :
+  'a -> 'b ->
+  ('a -> 'b -> 'ret, 'a -> 'b -> unit, 'ret) args
+
+(** [apply f l] applies a n-ary function [f] to the arguments from [l] *)
+val apply :
+  ('ar -> 'row) -> ('ar -> 'row, 'ar -> 'urow, 'ret) args ->
+  'ret
+
+(** GADT for function types.
+
```

Comment:
> Could you detail a bit more this comment? We want to know what is the exact problem you are solving with that GADT.

@yurug OK I'll try to add another sentence to better document that,
but note that the two GADTs of this file were not designed or implemented by me: they did exist before that PR in `Test_lib`, cf. https://github.com/ocaml-sf/learn-ocaml/blob/0.10/src/grader/test_lib.mli#L1052-L1069 

To sum this PR consists essentially in:

1. moving these two GADTs and the related functions from `Test_lib` to `Fun_ty`
2. adding a ppx extension point so that one can automatically build some term of type `(_, _, _) fun_ty` by writing `[%funty: bool -> char -> int]` instead of being forced to write something less legible such as `arg_ty [%ty: bool] (last_ty [%ty: char] [%ty: int])`.

Reply:
> Could you detail a bit more this comment? We want to know what is the exact problem you are solving with that GADT.

This should be addressed with my latest commit, cf. https://github.com/ocaml-sf/learn-ocaml/pull/302#discussion_r314510444